pkg/block/mdfetcher.go

```diff
@@ -0,0 +1,103 @@
+package block
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+type MetadataFetcher struct {
+	cachedBlockMDs map[ulid.ULID]*metadata.Meta
+	rwLock         *sync.RWMutex
+
+	concurrencyLimit int
+	logger           log.Logger
+	bkt              objstore.Bucket
+}
+
+func NewMetadataFetcher(lg log.Logger, concurrencyLimit int, bkt objstore.Bucket) *MetadataFetcher {
+	return &MetadataFetcher{
+		logger:           lg,
+		concurrencyLimit: concurrencyLimit,
+		rwLock:           &sync.RWMutex{},
+		cachedBlockMDs:   make(map[ulid.ULID]*metadata.Meta),
+		bkt:              bkt,
+	}
+}
+
+// Fetch wil return the mapping from block ID to its metadata that are currently stored in the bucket passed in.
+// If the corresponding value for a block ID key is nil, no valid metadta is retrieved successfully.
+func (f *MetadataFetcher) Fetch(ctx context.Context) (map[ulid.ULID]*metadata.Meta, error) {
```

Comment:
Why fetcher and not e.g `Syncer`? Especially if we maintain `cachedBlockMDs` for efficiency?

Reply:
Let's mention if it's goroutine safe  (: It looks like it is, the question is ... should it be?

I think it would be better if you would not make `Fetch/Sync` goroutine safe. First of all, I don't think it is needed. Secondly, with this, we can be more efficient in terms of memory. We can allow to reuse exactly the same `map[ulid.ULID]*metadata.Meta` (pass as an argument and return updated one). This might save some allocations as there can be millions of objects potentially in the bucket. 

What do you think? With this I am not sure if the struct is needed, maybe just Function is enough? Don't have a strong opinion.

Reply:
Be aware of this https://github.com/thanos-io/thanos/pull/1408 that changes this code as well (somehow we need to sync both PRs) cc @povilasv

Reply:
I am happy to rename to syncer if you feel strongly about it. I am alright with both. I chose fetch as I felt it is more about fetching the metadata from remote source. Keeping a cache is just an internal optimisation and might become an issue and get removed if the bucket contains too many objects.

Good point about about the memory allocation. I have a different perspective. I don't think it would add that much to memory as it is using metadata pointers right now. And if we were to reuse the same map, there would still have to be a map to keep what metadata are in the bucket in order to remove those cached but somehow-removed-from-remote-source ones. So the saving won't be that much. Besides, I think streaming metadata might be a better solution to scale with bigger number of objects in the bucket. What do you think?

Reply:
No strong feeling to the naming. It really depends on what exactly interface we are going to apply as you are exploring streaming suggestion now (: So then `ForeachMeta` might make sense more?

>  I think streaming metadata might be a better solution to scale with bigger number of objects in the bucket. What do you think?

Definitely, however, this means we would need to revert to any of Iter concepts like

* passing a function to invoke for each meta
* `Iter` interface: 
```
// Iterator is a simple iterator that can only get the next value.
type Iterator interface {
	At() metadata.Meta)
	Err() error
	Next() bool
}
```

Reply:
I think we don't have to use stream for now unless the OOM issue is already big and urgent. I think we can keep this signature which meshes pretty well with the current usage and we can use this as the foundation to identify and make changes when we are ready to explore other options like streaming.

Reply:
Yea, I think I am happy with this. We already keep `cached block MD` anyway.

Reply:
fetcher might be fine then